text/3720-labeled-match.md

```diff
@@ -0,0 +1,1088 @@
+
+Labeled match solves both the ergonomics issue and makes reliably generating efficient code much easier:
+
+```rust
+'top: match a {
+    1 => {
+        i += 1;
+        continue 'top 2;
+    }
+    2 => {
+        i += 1;
+        break 'top;
+    }
+    _ => break 'top,
+}
+
+// or even
+
+'top: match a {
+    1 => {
+        i += 1;
+        continue 'top 2;
+    }
+    2 => i += 1,
+    _ => {}
+}
+```
+
+One could argue that the inability to directly translate switch fallthrough into rust is an instance of the [XY problem](https://xyproblem.info/), but many parsers, interpreters and other state machines just rely on this kind of control flow.
+
+A niche, but very valuable use case is [c2rust](https://github.com/immunant/c2rust), a tool that automatically translates C to rust. In many cases, a C `switch` cannot be automatically translated to a rust `match` due to implicit fallthroughs, so the translation produces an abomination of labeled blocks and loops: correct, but hard to reason about. Being able to lower such control flow, in most cases, to a labeled match greatly improves both the readability of the generated code, speeding up the porting process.
+
+Many other parser, decoder and other lowlevel crates will similarly benefit from the ergonomics of labeled match.
+
+## Code generation
+
+State machines (parsers, interpreters, ect) can be written as a loop containing a match on the current state. The match picks the branch that belongs to the current state, some logic is performed, the state is updated, and eventually control flow jumps back to the top of the loop, branching to the next state.
+
+```rust
+loop {
+    match state {
+        A => {
+            // <perform work>
+
+            state = B;
+        }
+        B => {
+            // ...
+        }
+        // ...
+    }
+}
+```
+
+While this is a natural way to express a state machine, it is well-known that when translated to machine code in a straightforward way, this approach is inefficient on modern CPUs:
+
+- The match is an unpredictable branch, causing many branch misses. Reducing the number of branch misses is crucial for good performance on modern hardware.
+- The "loop + match" approach contains control flow paths (so, sequences of branches) that will never be taken in practice. The stack can be smaller if the control flow paths are known more precisely.
+
+By providing the compiler with more knowlege about what state transitions actually exists (i.e. what other states can follow a particular state), we get major performance improvements. A proof of concept implementation of labeled match shows considerable performance gains versus current recommended workarounds in real-world scenarios:
```

Comment:
If you are going to give benchmark numbers, please include the code you tested in a link.

Reply:
fixed now, but for completeness the instructions are here https://gist.github.com/folkertdev/977183fb706b7693863bd7f358578292